src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -56,6 +56,57 @@ class SessionWrapperStack {
     SessionWrapper* m_back = nullptr;
 };
 
+/// Register a function whose invocation can be triggered repeatedly.
+///
+/// While the function is always executed by the event loop thread, the
+/// triggering of its execution can be done by any thread.
+///
+/// The function is guaranteed to not be called after the Trigger object is
+/// destroyed.
+///
+/// Note that even though the trigger() function is thread-safe, the Trigger
+/// object, as a whole, is not. In particular, construction and destruction must
+/// not be considered thread-safe.
+class Trigger : public std::enable_shared_from_this<Trigger> {
+public:
+    Trigger(network::Service* service, std::function<void()>&& handler);
```

Comment:
Also, update handler to be a `SyncSocketProvider::FunctionHandler` so it matches the other function handler signatures we will be moving to.
You can call the handler with this inside Trigger::trigger()
```
self->handler(Status::OK());
```